bevy_lint/src/lints/borrow_of_reborrowable.rs

```diff
@@ -0,0 +1,224 @@
+                let TyKind::Ref(region, ty, Mutability::Mut) = arg_ty.kind() else {
+                    // We only care about `&mut` parameters
+                    continue;
+                };
+
+                let peeled_ty = ty.peel_refs();
+
+                let Some(reborrowable) = Reborrowable::try_from_ty(cx, peeled_ty) else {
+                    // The type is not one of our known re-borrowable types
+                    continue;
+                };
+
+                let is_output_bound_to_arg = fn_sig
+                    .output()
+                    .visit_with(&mut ContainsRegion(*region))
+                    .is_break();
+
+                if is_output_bound_to_arg {
+                    // We don't want to suggest re-borrowing if the return type's
+                    // lifetime is bound to the argument's reference.
+                    // This is because it's impossible to convert something like:
+                    // `for<'a> (&'a mut Commands<'_, '_>) -> EntityCommands<'a>`
+                    // to something like:
+                    // `for<'a> (Commands<'_, '_>) -> EntityCommands<'a>`
+                    // without getting: `error[E0515]: cannot return value referencing function parameter `commands``
+                    continue;
+                }
+
+                let arg_ident = arg_names[arg_index];
+                let span = decl.inputs[arg_index].span.to(arg_ident.span);
+
+                span_lint_and_sugg(
+                    cx,
+                    reborrowable.lint(),
+                    span,
+                    reborrowable.message(),
+                    reborrowable.help(),
+                    reborrowable.suggest(arg_ident, peeled_ty.to_string()),
+                    // Not machine-applicable since the function body may need to
+                    // also be updated to account for the removed ref
+                    Applicability::MaybeIncorrect,
+                );
+            }
+        });
+    }
+}
+
+#[derive(Debug)]
+enum Reborrowable {
+    Commands,
+    // Deferred,
+    // DeferredWorld,
+    EntityCommands,
+    // EntityMut,
+    // EntityMutExcept,
+    // FilteredEntityMut,
+    // FilteredResourcesMut,
+    // Mut,
+    // MutUntyped,
+    NonSendMut,
+    // PtrMut,
+    Query,
+    // QueryIterationCursor,
+    ResMut,
+}
+
+impl Reborrowable {
+    fn try_from_ty<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> Option<Self> {
+        if match_type(cx, ty, &crate::paths::COMMANDS) {
+            Some(Self::Commands)
+        } else if match_type(cx, ty, &crate::paths::ENTITY_COMMANDS) {
+            Some(Self::EntityCommands)
+        } else if match_type(cx, ty, &crate::paths::QUERY) {
+            Some(Self::Query)
+        } else if match_type(cx, ty, &crate::paths::RES_MUT) {
+            Some(Self::ResMut)
+        } else if match_type(cx, ty, &crate::paths::NON_SEND_MUT) {
+            Some(Self::NonSendMut)
+        } else {
+            None
+        }
```

Comment:
```suggestion
        const PATH_MAP: &[(&[&str], Reborrowable)] = &[
            (&crate::paths::COMMANDS, Reborrowable::Commands),
            (&crate::paths::ENTITY_COMMANDS, Reborrowable::EntityCommands),
            (&crate::paths::QUERY, Reborrowable::Query),
            (&crate::paths::RES_MUT, Reborrowable::ResMut),
            (&crate::paths::NON_SEND_MUT, Reborrowable::NonSendMut),
        ];

        for &(path, reborrowable) in PATH_MAP {
            if match_type(cx, ty, path) {
                return Some(reborrowable);
            }
        }

        None
```

I refactored this to be a little more compact, and a little easier to extend in the future. You do need to derive `Copy` and `Clone` on `Reborrowable`, but that's find since it's only 1 byte large.

Reply:
I like it!